### PR TITLE
LessThan circuits and padding

### PR DESCRIPTION
**Pull request by Antonio95**
This PR adds two example circuits to verify `a < b`, where `a` is a private witness held by the prover and `b` is a public constant fixed at set-up. The two circuits rely on a fixed bit length that `a` is supposed to fit in (when regarded as an unsigned integer, even though in reality it lies in some finite field).
- The circuit `LessThanCircuitSafe` performs the aforementioned bit length verification.
- The circuit `LessThanCircuitUnsafe` does not. This leads to constraint savings when the circuit is used as a component in a larger circuit which already ensures the bit length meets the required bound.

In the course of writing the examples, we encountered the issue that the existing Spartan machinery did not pad the matrices to power-of-two row and column lengths, as necessitated by the proof system. This caused compilation errors in our examples. Since it seemed like a good idea for this need for padding to be abstracted away from the user, we have implemented padding in the Spartan internals - that is the other significant contribution in this PR. Note that a `R1CSShape` method `.pad()` already existed [here](https://github.com/microsoft/Spartan2/blob/main/src/r1cs.rs#L298), but after some experimentation we convinced ourselves that it could not be used to pad the circuit in the cases where it was necessary.

Other minor changes:
-  The aforementioned circuits are accompanied by meaningful tests (in the example code) that try edge cases and also showcase the different execution outcomes of the safe and unsafe variants for some inputs.
- We have removed the `<F>` generic on the existing `CubicCircuit` example, which in turn allowed us to remove the `PhantomData` therein and didn't require any further code modifications.
- A section on how to run Rust examples (which the new circuits are) has been added to the README.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -178,8 +176,7 @@ mod tests {
     let circuit = CubicCircuit::default();
 
     // produce keys
-    let (pk, vk) =
-      SNARK::<G, S, CubicCircuit<<G as Group>::Scalar>>::setup(circuit.clone()).unwrap();
+    let (pk, vk) = SNARK::<G, S, CubicCircuit>::setup(circuit.clone()).unwrap();
```
Comment (mmagician) [inline src/lib.rs:179]: that's nice

File: src/spartan/snark.rs [hunk 2]
```diff
@@ -101,6 +101,26 @@ impl<G: Group, EE: EvaluationEngineTrait<G>> RelaxedR1CSSNARKTrait<G> for Relaxe
...
+    let num_constraints = cs.num_constraints();
+
+    (num_constraints..num_constraints.next_power_of_two()).for_each(|i| {
+      cs.enforce(
+        || format!("padding_constraint_{i}"),
+        |lc| lc,
+        |lc| lc,
+        |lc| lc,
+      )
+    });
```
Comment (mmagician) [inline src/spartan/snark.rs:115]: I don't understand what this code does. Is this a dummy constraint?

File: src/spartan/snark.rs [hunk 3]
```diff
@@ -121,6 +141,14 @@ impl<G: Group, EE: EvaluationEngineTrait<G>> RelaxedR1CSSNARKTrait<G> for Relaxe
...
     let _ = circuit.synthesize(&mut cs);
 
+    // Padding variables
+    let num_vars = cs.aux_slice().len();
+
+    (num_vars..num_vars.next_power_of_two()).for_each(|i| {
+      cs.alloc(|| format!("padding_var_{i}"), || Ok(G::Scalar::ZERO))
+        .unwrap();
+    });
+
```
Comment (mmagician) [inline src/spartan/snark.rs:151]: Why does the prover only pad variables and not constraints like in the setup function?

**Review by mmagician — COMMENTED** [review]
Also, shouldn't there be an analogous modification for the verifier?

File: src/spartan/snark.rs (hunk 2 above)
Comment (Antonio95) [reply src/spartan/snark.rs:115]: Precisely. Recall that in an R1CS the number of constraints corresponds to the number of rows. The issue with the previous implementation was that if that number was not a power of 2, the compiler would break (funnily enough, the only example they had, the `CubicCircuit`, had exactly 4 constraints by nature - although I think I have a way to do it with 3). So we add "empty constraints" where all coefficients are zero - and yes, that's what that slightly odd syntax is doing. We also needed to add zero-valued variables to pad the number of columns to a power of 2.

File: src/spartan/snark.rs (hunk 3 above)
Comment (Antonio95) [reply src/spartan/snark.rs:151]: The variable that this function receives representing the circuit, `circuit`, is produced by `setup` and therefore had its variables and constraints padded elsewhere. The `cs` here is only a `SatisfyingAssignment` (basically the vector (1 || witness || instance)), and so we only need to pad it to the right (one-dimensional) length.

**Comment by Antonio95** [comment]
> Also, shouldn't there be an analogous modification for the verifier?

RIght now, there are two padding-related modifications:
- One on `setup` which adds constraint and variable padding: this method affects both the prover and the verifier. 
- One on `prove` which padds the (1 || witness || instance) vector (cf. above).

The verfier doesn't receive the latter vector, and their circuit is padded during `setup`. So, as far as I can tell, nothing is needed on the verifier side (cf the verification code [here](https://github.com/HungryCatsStudio/Spartan2/blob/less_than/src/spartan/snark.rs#L407)). The tests more or less bear this out, since they broke with either of the aforementioned modifications and they work with the current code.

File: examples/less_than.rs [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+
+// Range check: constrains input < `bound`. The bound must fit into
+// `num_bits` bits (this is asserted in the circuit constructor).
+// Important: it must be checked elsewhere (in an overarching circuit) that the
+// input fits into `num_bits` bits - this is NOT constrained by this circuit
+// in order to avoid duplications (hence "unsafe"). Cf. LessThanCircuitSafe for
+// a safe version.
+#[derive(Clone, Debug)]
+struct LessThanCircuitUnsafe<F: PrimeField> {
+  bound: F, // Will be a constant in the constraits, not a variable
```
Comment (mmagician) [inline examples/less_than.rs:72]: I think that a more natural representation of `bound` is as unsigned integer, rather than a field element. For most cases a `u8` would suffice, but with `u16` you're definitely good.

File: examples/less_than.rs [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+}
+
+impl<F: PrimeField + PrimeFieldBits> Circuit<F> for LessThanCircuitUnsafe<F> {
+  fn synthesize<CS: ConstraintSystem<F>>(self, cs: &mut CS) -> Result<(), SynthesisError> {
+    assert!(F::NUM_BITS > self.num_bits as u32 + 1);
+
+    let input = AllocatedNum::alloc(cs.namespace(|| "input"), || Ok(self.input))?;
+
+    let shifted_diff = AllocatedNum::alloc(cs.namespace(|| "shifted_diff"), || {
+      Ok(self.input + F::from(1 << self.num_bits) - self.bound)
```
Comment (mmagician) [inline examples/less_than.rs:94]: Then this would change to:
```suggestion
      Ok(self.input + F::from(1 << self.num_bits - self.bound))
```

File: src/spartan/snark.rs (hunk 2 above)
Comment (mmagician) [reply src/spartan/snark.rs:115]: Alright, makes sense on a high level. I'm not as familiar with the internals as you at this point, let's go upstream with this and see.

File: examples/less_than.rs (hunk 4 above)
Comment (Antonio95) [reply examples/less_than.rs:72]: Yes, in fact originally both the input and the bound were `u64`s (cf. [here](
https://github.com/HungryCatsStudio/Spartan2/blob/b7bf232d65fdb246af7f71af984d82e7c28031f6/examples/less_than.rs#L59-L60)). Then we needed to change the input to `F` because we realised only then can one (in general) show examples where the safe version detects cheating which the unsafe version doesn't. And, since the input became an `F`, it made sense to us to have the bound as an `F` as well. This change of the bound type in fact came at the cost of programming of [a small extra function](https://github.com/HungryCatsStudio/Spartan2/blob/antonio-cesar/less-than/examples/less_than.rs#L53) by @Cesar199999 because an original constructor check could no longer be applied (original check [here](https://github.com/HungryCatsStudio/Spartan2/blob/b7bf232d65fdb246af7f71af984d82e7c28031f6/examples/less_than.rs#L66), new check [here](https://github.com/HungryCatsStudio/Spartan2/blob/antonio-cesar/less-than/examples/less_than.rs#L78)).

I think having the bound as `F` is more general: one can always use `F::from` if one has an unsigned integer bound, but the converse cannot be done if the bound is passed to the constructor as a `u_something` and one has an `F`. Furthermore, most fields we use are larger than even `u128`, so with an `F` one can express more bounds. Note that this doesn't cause extra constraints, since the number of constraints is dictated by the user-defined variable `num_bits`. This is an example for the library, so making it more general than our specific needs doesn't seem pointless.

Having said all of that, we can totally change it to a `u_something` if you prefer. I'd say `u16` is unnecessarily small, maybe `u64` (as originally) better? That's also the type `F::from` wants. If we make this change, we could do away with `get_msb_index` (going back to the original check) and add the check that the new unsigned bound fits into the field.

Comment (mmagician) [reply examples/less_than.rs:72]: Oh I thought this is meant to represent the bound on the number of bits, rather than the bound itself, hence I proposed `u16`.

Comment (mmagician) [reply examples/less_than.rs:72]: Kind of like the `shift_bound` in the r1cs exercises: https://github.com/rdi-berkeley/zkp-mooc-lab/blob/main/circuits/float_add.circom#L256

Comment (Antonio95) [reply examples/less_than.rs:72]: Not quite, so: these chips implement the check `input < bound`, where the LHS is a witness and the RHS is a constant. So, in [here](https://github.com/HungryCatsStudio/Spartan2/blob/eb2843e26ff653d074ecaef5b1a47876e832f964/examples/less_than.rs#L124-L126), `input` and `bound` are as I just described, and `num_bits` represents what input size should be supported. For instance, if `num_bits = 32`, then inputs between 0 and 2^32 - 1 are supported. This is of course different from the number of bits of the bound: in the previous example, one could have e.g. `bound = 2024`, which is 11 bits long (but we might want to compare inputs between 0 and 2^32 - 1).

Comment (Antonio95) [reply examples/less_than.rs:72]: I have just updated the documentation, which described the circuit as a range check, since that might be misleading (i.p. it sounds like the bound is of the form `2^n`, which might have confused you @mmagician)

Comment (mmagician) [reply examples/less_than.rs:72]: Ah okey, yes makes sense.

**Review by mmagician — APPROVED** [review]
LGTM, PR away to upstream!
